File: src/Users/Application/Login/LoginHandler.php
<?php

declare(strict_types=1);

namespace App\Users\Application\Login;

use App\Shared\Domain\Exceptions\CustomException;
use App\Shared\Domain\JwtManagerInterface;
use App\Shared\Domain\JwtPayload;
use App\Users\Domain\User;
use App\Users\Domain\UserEmail;
use App\Users\Domain\UserFinder\UsersFinder;

class LoginHandler
{
    public function __construct(
        private readonly UsersFinder         $usersFinder,
        private readonly JwtManagerInterface $jwtManager,
    )
    {
    }

    /**
     * @throws CustomException
     */
    public function execute(LoginCommand $command): LoginResult
    {
        $user = $this->usersFinder->getUserByEmail(UserEmail::fromValue($command->email));
        $this->checkPasswordOrThrow($user, $command->password);
        $token = $this->jwtManager->encode(new JwtPayload($user->getId()->value));
        return new LoginResult($token);
    }

    /**
     * @throws CustomException
     */
    private function checkPasswordOrThrow(User $user, string $password): void
    {
        if (!$user->getPassword()->compare($password)) {
            throw new CustomException('Usuario o contraseña incorrecto', 401);
        }
    }
}

File: src/main/kotlin/com/zaleort/second_brain/Users/Application/UseCases/Login/LoginHandler.kt
package com.zaleort.second_brain.Users.Application.UseCases.Login

import com.zaleort.second_brain.Shared.Domain.JwtManager.JwtPayload
import com.zaleort.second_brain.Shared.Infrastructure.JwtManager.JwtManager
import com.zaleort.second_brain.Users.Domain.User.Exceptions.InvalidCredentialsException
import com.zaleort.second_brain.Users.Domain.User.UserEmail
import com.zaleort.second_brain.Users.Domain.User.UserRepositoryInterface
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service

@Service
class LoginHandler(
    val repository: UserRepositoryInterface,
    val jwtManager: JwtManager,
) {
    fun execute(command: LoginCommand): LoginDTO {
        val userEmail = UserEmail(command.email)
        val user = this.repository.findByEmail(userEmail)
            ?: throw InvalidCredentialsException("Invalid user or password")

        val encoder = BCryptPasswordEncoder()
        println(user.password.value)
        println(command.password)
        if (!encoder.matches(command.password, user.password.value)) {
            throw InvalidCredentialsException("Invalid user or password")
        }

        val token = jwtManager.encode(JwtPayload(user.id.value))
        return LoginDTO(token)
    }
}